library/alloc/src/macros.rs

```diff
@@ -101,9 +101,21 @@ macro_rules! vec {
 /// ```
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[allow_internal_unstable(fmt_as_str)]
 macro_rules! format {
     ($($arg:tt)*) => {{
-        let res = $crate::fmt::format($crate::__export::format_args!($($arg)*));
-        res
+        // NOTE: `format_args!` borrows from temporaries. This means that
+        // `match` is necessary to extend the lifetime of the temporaries until after
+        // the `Arguments` is no longer used. The same pattern is used
+        // inside `format_args!` itself.
+        let r = match $crate::__export::format_args!($($arg)*) {
+            // FIXME: We hope that constant propagation will make LLVM optimize out
+            // this match.
+            args => match args.as_str() {
+                Some(s) => $crate::borrow::ToOwned::to_owned(s),
+                None => $crate::fmt::format(args),
+            }
+        };
```

Comment:
I still think this match should be in `crate::fmt::format` so we're not special casing `format!`.

Reply:
`alloc::fmt::format` is not inline, so this match would not be folded away:

https://github.com/rust-lang/rust/blob/65d071eeb52eba78896dd09b95a80c647ef150e9/library/alloc/src/fmt.rs#L572-L574

Reply:
the problem is that const prop doesn't work when this check is in `fmt::format`, we would just have early returning and more branch for this case.

Reply:
I see :/ maybe it makes sense to do the same thing for other macros using `format_args!` then?

Reply:
I did the similar idea you proposed in https://github.com/rust-lang/rust/pull/75358, however the result is not reliable
and mostly a regression.

Reply:
That's unfortunate, if there's a perf regression no point in adding it elsewhere.

Reply:
To clarify: In https://github.com/rust-lang/rust/pull/75358, I used the match `Arguments::as_str` in the method directly.
While this PR is adding check in macro.